garage/misc/git_handler.py

```diff
@@ -0,0 +1,336 @@
+        full_tag_name = "refs/tags/" + tag_name
+        return self.get_sha(full_tag_name)
+
+    def reset(self, reference="HEAD"):
+        """Reset the working directory to the indicated reference."""
+        self.repo.git.reset("--hard", reference)
+
+    def get_sha(self, reference="HEAD"):
+        """Return the 40 character SHA of the reference.
+
+        The reference can be a local branch ("refs/heads/<branch_name>"), a
+        remote branch ("refs/remotes/<branch_name>"), a tag
+        ("refs/tags/<tag_name>"), HEAD or a SHA.
+        Passing a SHA as reference comes useful to verify the SHA is valid.
+        HEAD is just a reference to the commit where the working directory is
+        currently checked out.
+
+        If the reference is not valid, the exception sh.ErrorReturnCode_128 is
+        thrown.
+        """
+        try:
+            commit = self.repo.rev_parse(reference)
+        except (BadName, ValueError):
+            print(
+                colored("The reference %s does not exist", "red") % reference)
+            raise
+        return str(commit)
+
+    def create_stash(self):
+        """Stashes the local changes that have not been staged.
+
+        The stash is not stored anywhere in the ref namespace so the user does
+        not have remove it manually later.
+
+        Returns
+        -------
+        The 40 character SHA if local changes were found, or an empty string
+        otherwise.
+
+        """
+        stash_sha = ""
+        try:
+            self.repo.git.diff(exit_code=True)
+        except GitCommandError:
+            # Exit code 1 means there are differences
+            stash_sha = self.repo.git.stash("create")
+            pass
+        return stash_sha
+
+    def get_current_branch_name(self):
+        """Return the name of the checked out branch.
+
+        If the HEAD is in detached state, an empty string is returned.
+
+        Precondition
+        ------------
+        - No branch (local or remote) in the repository is named "HEAD".
+
+        """
+        curr_branch_name = self.repo.git.rev_parse(
+            self.repo.head, abbrev_ref=True)
+        if curr_branch_name == "HEAD":
+            curr_branch_name = ""
+        return curr_branch_name
+
+    def run_experiment_on_target(self, target, exp_args):
+        """Run a experiment on the indicated commit defined by target.
+
+        After the target is verified, if there are any non-staged changes in
+        the working directory, they're stashed, the target commit is checked
+        out, the experiment is executed and the working directory is restored
+        at the end, including the stashed changes.
+
+        Parameters
+        ----------
+        - target: a key-value string, where the key indicates the type of
+            target (tag, branch, sha), and the value the reference for the
+            corresponding type. For example:
+                sha: 8d54s23a
+                branch: my_branch
```

Comment:
git should be able to disambiguate these without knowing what kind of ref it is

Reply:
That's true if all of them have different name, but what if they don't. For example:
```
$ git branch my_feature
$ git tag my_feature
$ git rev-parse my_feature
warning: refname 'my_feature' is ambiguous.
4421c2a81debf46079793eee05154f8a73ab3faa
```